building: prevent pyqtgraph.canvas from crashing binary dependency analysis

rokm opened the pull request:
When the isolated subprocess crashes while importing packages prior to binary dependency analysis (to infer dynamic DLL search paths on Windows), include the list of planned package imports in the exception message.

This improves our chances of reproducing the issue when the crash is related to the order in which packages are imported.

rokm committed:
building: extend error message for crashed subprocess during binary dep. analysis

When the isolated subprocess crashes while importing packages prior
to binary dependency analysis (to infer dynamic DLL search paths
on Windows), include the list of planned package imports in the
exception message.

This improves our chances of reproducing the issue when the crash
is related to the order in which packages are imported.

bwoodsend commented:
Perhaps, if we're augmenting that specific error anyway, we could word it in a way that explains to the user that they're seeing this error because they have a package which crashes on import in their environment and then they might be able to fix their issue on their own? Obviously it's an ambitious *if*.

Consider this PR approved if you don't think the above is likely to help.

rokm commented:
I think the problem is that we have two kinds of crashes; one is where importing a package per-se causes the crash (= broken package), and the other is related to order of imports. I.e., a package is imported before another package, whereas typical import order in user program is inverse; as an example, user might typically import `PyQt6` (or bindings of their choice) before importing `pyqtgraph`.

We've seen the first kind of crash with `pyqtgraph.canvas` in https://github.com/pyinstaller/pyinstaller/issues/7991 while trying to collect submodules from it as part of  `collect_submodules`. In contrast, the crash in #8322 happens during package imports prior to binary dependency analysis (so the import during the hook stage must have succeeded), which makes me suspect that it is related to order of imports. By dumping the list of packages that we were trying to import, I could try to reproduce the crash by simply following the same list.

(And I suppose at the end of the day, to deal with such crashes, we will probably have to exempt problematic packages from being imported either way - but I would prefer to understand the exact failure mode for each of them).

rokm committed:
building: improve order of package imports prior to binary dep. analysis

When importing packages for tracking of dynamic library search
path modifications prior to binary dependency analysis, ensure
that Qt bindings packages are always imported first.

This prevents other packages that support multiple Qt bindings
from trying to auto-select and import an available Qt bindings
package that was excluded from the analysis (either manually
or automatically).

But just to be sure that no Qt bindings package except the
collected one(s) can be imported in the isolated subprocess,
explicitly suppress them by setting their `sys.module`
entries to `None` in the setup/preparation stage.

rokm committed:
building: avoid importing pyqtgraph.canvas prior to binary dep. analysis

Add `pyqtgraph.canvas` to the list of packages whose import
should be explicitly suppressed when importing packages for
tracking of dynamic library search path modifications
prior to binary dependency analysis.

The `pyqtgraph.canvas` is known to crash python interpreter
in certain circumstances, and while the issue has been fixed
upstream, we need to work around the issue for older releases
(<= 0.13.3).

rokm commented:
As per https://github.com/pyinstaller/pyinstaller/issues/8322#issuecomment-1966391104, this now attempts to ensure that collected Qt bindings (if any) are imported first, so that other packages that auto-detect bindings by looking at what is already in `sys.modules` select the correct bindings (and do not try to do their own try-import autodetection). Just to be safe, we now also suppress all non-collected Qt bindings, as those might still be available in environment (and while they might have been explicitly excluded from analysis via `--exclude` or #8329) and some other package might try to import them (and succeed in that).

Lastly, suppress import of `pyqtgraph.canvas` due to its tendency to crash python interpreter (see https://github.com/pyinstaller/pyinstaller/issues/8322#issuecomment-1971971486). The underlying bug was fixed in `pyqtgraph`, but we need to work around for current and older releases (<= 0.13.3).
